internal/cmd/oohelperd/internal/nwcth/measure.go

```diff
@@ -0,0 +1,281 @@
+
+	CtrlHTTPMeasurement = nwebconnectivity.ControlHTTP
+
+	CtrlH3Measurement = nwebconnectivity.ControlH3
+
+	CtrlTLSMeasurement = nwebconnectivity.ControlTLSHandshake
+
+	CtrlHTTPRequest = nwebconnectivity.ControlHTTPRequest
+)
+
+// NextLocationInfo contains the redirected location as well as the request object which forwards most headers of the initial request.
+// This forwarded request is generated by the http.Client and
+type NextLocationInfo struct {
+	jar             *cookiejar.Jar `json:"-"`
+	location        string         `json:"-"`
+	httpRedirectReq *http.Request  `json:"-"`
+}
+
+type MeasureURLResult struct {
+	CtrlURLMeasurement *CtrlURLMeasurement `json:"-"`
+	redirectedReqs     []*CtrlRequest      `json:"-"`
+	h3Reqs             []*CtrlRequest      `json:"-"`
+}
+
+type MeasureEndpointResult struct {
+	CtrlEndpoint CtrlEndpointMeasurement
+	httpRedirect *NextLocationInfo
+	h3Location   string
+}
+
+func Measure(ctx context.Context, creq *CtrlRequest) (*CtrlResponse, error) {
+	var cresp = &CtrlResponse{URLMeasurements: []*CtrlURLMeasurement{}}
+
+	redirected := make(map[string]bool, 100)
+
+	urlM, err := MeasureURL(ctx, creq, cresp, redirected)
+	if err != nil {
+		return nil, err
+	}
+	cresp.URLMeasurements = append(cresp.URLMeasurements, urlM.CtrlURLMeasurement)
+
+	n := 0
+	nextRequests := append(urlM.redirectedReqs, urlM.h3Reqs...)
+	for len(nextRequests) > n {
+		req := nextRequests[n]
+		n += 1
+		if len(redirected) == 20 {
+			// stop after 20 redirects
+			break
+		}
+		if _, ok := redirected[req.HTTPRequest]; ok {
+			continue
+		}
+		redirected[req.HTTPRequest] = true
+		urlM, err := MeasureURL(ctx, req, cresp, redirected)
+		if err != nil {
+			return nil, err
+		}
+		cresp.URLMeasurements = append(cresp.URLMeasurements, urlM.CtrlURLMeasurement)
+		nextRequests = append(nextRequests, urlM.redirectedReqs...)
+	}
+
+	return cresp, nil
+}
+
+// Measure performs the measurement described by the request and
+// returns the corresponding response or an error.
+func MeasureURL(ctx context.Context, creq *CtrlRequest, cresp *CtrlResponse, redirected map[string]bool) (*MeasureURLResult, error) {
+	// parse input for correctness
+	URL, err := url.Parse(creq.HTTPRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// create URLMeasurement struct
+	urlMeasurement := &CtrlURLMeasurement{
+		URL:       URL.String(),
+		DNS:       nil,
+		Endpoints: []CtrlEndpointMeasurement{},
+	}
+
+	// dns: start
+	dns := DNSDo(ctx, &DNSConfig{
+		Domain: URL.Hostname(),
+	})
+
+	urlMeasurement.DNS = &dns
+
+	enpnts := getEndpoints(dns.Addrs, URL)
+	addrs := mergeEndpoints(enpnts, creq.TCPConnect)
+
+	if len(addrs) == 0 {
+		return nil, errors.New("no valid IP address to measure")
+	}
+
+	wg := new(sync.WaitGroup)
+	out := make(chan *MeasureEndpointResult, len(enpnts))
+	for _, endpoint := range enpnts {
+		wg.Add(1)
+		go MeasureEndpoint(ctx, creq, URL, endpoint, wg, out)
+	}
+	wg.Wait()
+	close(out)
```

Comment:
It seems to me closing this channel has no effect, am I missing something?

Reply:
A few lines underneath, the channel is iterated. `for m := range out {...}` 
I'm pretty sure that for the loop to terminate, the channel has to be closed.